bundles/org.openhab.binding.salus/src/main/java/org/openhab/binding/salus/internal/rest/SalusApi.java

```diff
@@ -0,0 +1,219 @@
+ */
+@NonNullByDefault
+public class SalusApi {
+    private static final int MAX_TIMES = 3;
+    private final Logger logger;
+    private final String username;
+    private final char[] password;
+    private final String baseUrl;
+    private final RestClient restClient;
+    private final GsonMapper mapper;
+    @Nullable
+    private AuthToken authToken;
+    @Nullable
+    private LocalDateTime authTokenExpireTime;
+    private final Clock clock;
+
+    public SalusApi(String username, char[] password, String baseUrl, RestClient restClient, GsonMapper mapper,
+            Clock clock) {
+        this.username = username;
+        this.password = password;
+        this.baseUrl = removeTrailingSlash(baseUrl);
+        this.restClient = restClient;
+        this.mapper = mapper;
+        this.clock = clock;
+        // thanks to this, logger will always inform for which rest client it's doing the job
+        // it's helpful when more than one SalusApi exists
+        logger = LoggerFactory.getLogger(SalusApi.class.getName() + "[" + username.replace(".", "_") + "]");
+    }
+
+    public SalusApi(String username, char[] password, String baseUrl, RestClient restClient, GsonMapper mapper) {
+        this(username, password, baseUrl, restClient, mapper, Clock.systemDefaultZone());
+    }
+
+    private RestClient.Response<@Nullable String> get(String url, RestClient.Header header, int times) {
+        refreshAccessToken();
+        var response = restClient.get(url, authHeader());
+        if (response.statusCode() == 401) {
+            logger.info("Refreshing access token");
+            login(username, password);
+            if (times > MAX_TIMES) {
+                logger.debug("Could not refresh access token after {} times", MAX_TIMES);
+                return response;
+            }
+            return get(url, header, times + 1);
+        }
+        return response;
+    }
+
+    private RestClient.Response<@Nullable String> post(String url, RestClient.Content content, RestClient.Header header,
+            int times) {
+        refreshAccessToken();
+        var response = restClient.post(url, content, header);
+        if (response.statusCode() == 401) {
+            logger.debug("Refreshing access token");
+            login(username, password);
+            if (times > MAX_TIMES) {
+                logger.debug("Could not refresh access token after {} times", MAX_TIMES);
+                return response;
+            }
+            return post(url, content, header, times + 1);
+        }
+        return response;
+    }
+
+    private static String removeTrailingSlash(String str) {
+        if (str.endsWith("/")) {
+            return str.substring(0, str.length() - 1);
+        }
+        return str;
+    }
+
+    private void login(String username, char[] password) {
+        login(username, password, 1);
+    }
+
+    private void login(String username, char[] password, int times) {
```

Comment:
Can you add the 'throws HttpUnauthorizedException' to the method declaration including the other possible excpetions and also popegate those to the callers.

Reply:
I've removed HttpException.java completely